### DISCUSSION TITLE
Accessing last trial data outside of a jspsych variable

### DISCUSSION THREAD

**Discussion by MelissaJEvans** [jspsych/jsPsych#1343] [Q&A]
Hey! 

I'm currently trying to take what a participant enters as their response in a jsPsych-survey-text plugin to determine what trials they will see. I was trying to use jsPsych.data.getLastTrialData() to pull out their response, but then I couldn't then use that information to change a variable that exists outside of a jsPsych plugin variable. 
I gather I cannot use  jsPsych.data.getLastTrialData() in normal script, I need to use this fx within a jsPsych plugin, but can I manipulate a pre-existing variable within a jsPsych variable? 

Apologies if that doesn't make much sense, let me know if I need to clarify.

**Comment by jodeleeuw** [comment]
Hi @MelissaJEvans,

I think I need a little more information to understand what you are trying to do.

Are you using [dynamic parameters](https://www.jspsych.org/overview/trial/#dynamic-parameters) to set these variables?

**Reply by MelissaJEvans** [reply]
Yes!
I have a survey-text variable and I would like to take the response entered in the textbox to determine which set of stimuli participants see. To accomplish that, I want to be able to modify a global variable based on their input. 

What I did was establish a dynamic parameter like the example you linked to, but since I cannot use  jsPsych.data.getLastTrialData() outside of a jsPsych variable, I haven't been able to modify the global variable. 
(I'm relatively new to JavaScript so I may just be going about this the wrong way. Thanks for any advice!)

**Reply by jodeleeuw** [reply]
Got it, thanks for the clarification!

You can *modify* the global variable from inside the jsPsych functions. The problem is probably related to when you are trying to *read* the value of the global variable. If you have something like this...

```js

var global_variable = null;

var trial = {
	on_finish: function(data){
		global_variable = data.response
	}
}

if(global_variable !== null){

}

jsPsych.init({
	timeline: timeline
})
```

... then the value of `global_variable` will still be `null` when the code reaches the `if(...)` statement. This is because the experiment hasn't run yet. The experiment doesn't run until `jsPsych.init()` and so the `on_finish` function ends up running *after* the `if(...)` statement even though it is earlier in the written code.

The best solution here depends on the complexity of what you want to do with `global_variable`. If you are just choosing between a few different sets of stimuli, then one option is to use [conditional timelines](https://www.jspsych.org/overview/timeline/#conditional-timelines). But you'll need a separate timeline for each possible path, so that gets unwieldy if you have lots of possible paths.

Another alternative is to use dynamic parameters in the subsequent trials. So the trials themselves would be looking at the state of `global_variable` and changing their parameters based on its value. I do something sort of like this in my [tutorial on building an n-back task on YouTube](https://www.youtube.com/watch?v=fYKaKTvybCI&list=PLnfo1lBY1P2Mf_o6rV5wiqqn92Mw3UTGh&index=1).

**Reply by MelissaJEvans** [reply]
That text is quite like what I have! That makes sense, my global variable didn't seem to change by the time I got to the if statement, now I understand why. I'll look at those solutions, thanks!
